build: Suppress -Wdeprecated-copy warnings

hebasto opened the pull request:
Tomorrow, on Apr 23 the Ubuntu 20.04 release is expected. It packaged with Qt 5.12 LTS that has a nasty peculiarity to cause modern compilers, including Clang 10.0 and GCC 9.3, to emit spammy `-Wdeprecated-copy` warnings (#15822, #18419).

This PR suppress such warnings _temporarily_, until the [upstream is fixed](https://codereview.qt-project.org/c/qt/qtbase/+/272258).

Here are some affected systems (with system packages):
- Ubuntu 20.04 LTS + Qt 5.12.8 LTS + { Clang 10.0 | GCC 9.3 }
- Fedora 32 + Qt 5.13.2 + Clang 10.0

Reference: [QTBUG-75210](https://bugreports.qt.io/browse/QTBUG-75210)

Also see **fanquake**'s [comment](https://github.com/bitcoin/bitcoin/pull/18738#issuecomment-622956100).

hebasto committed:
build: Suppress -Wdeprecated-copy warnings

hebasto commented:
[**Clang 10.0**](https://packages.ubuntu.com/focal/clang) [Release Notes](https://releases.llvm.org/10.0.0/tools/clang/docs/ReleaseNotes.html#improvements-to-clang-s-diagnostics):
> - `-Wextra` now enables `-Wdeprecated-copy`. The warning deprecates move and copy constructors in classes where an explicit destructor is declared. This is for compatibility with GCC 9, and forward looking for a change that’s being considered for C++23. You can disable it with `-Wno-deprecated-copy`.

MarcoFalke commented:
I can't believe that qt is so horribly incompetent that they can't even run a compiler with warnings enabled on their header files.

MarcoFalke commented:
Is there a way to only add this flag when a file is compiled in the `src/qt` folder?

hebasto commented:
@MarcoFalke 
> Is there a way to only add this flag when a file is compiled in the `src/qt` folder?

Unfortunately, it does not help. Warnings are emitted from `/usr/include/x86_64-linux-gnu/qt5/QtCore/` and `/usr/include/x86_64-linux-gnu/qt5/QtWidgets/` dirs.

It seems @Empact's 422dcd5b4e2bedefc25f04741ff9718835551dc7 from #14920 or @vasild's c9376ae34cb9ab9e1634fefb080ecdbe33c2656b from #18149 could help, no?

MarcoFalke commented:
The warnings are emitted when compiling our `.cpp` files located in `src/qt`. See for example the full include trace:

```
In file included from qt/rpcconsole.cpp:9:
In file included from ./qt/rpcconsole.h:8:
In file included from ./qt/guiutil.h:12:
In file included from /usr/include/qt5/QtWidgets/QHeaderView:1:
In file included from /usr/include/qt5/QtWidgets/qheaderview.h:44:
In file included from /usr/include/qt5/QtWidgets/qabstractitemview.h:44:
In file included from /usr/include/qt5/QtWidgets/qabstractscrollarea.h:44:
In file included from /usr/include/qt5/QtWidgets/qframe.h:44:
In file included from /usr/include/qt5/QtWidgets/qwidget.h:45:
In file included from /usr/include/qt5/QtCore/qobject.h:47:
/usr/include/qt5/QtCore/qstring.h:1061:22: warning: definition of implicit copy constructor for 'QCharRef' is deprecated because it has a user-declared copy assignment operator [-Wdeprecated-copy]
    inline QCharRef &operator=(const QCharRef &c) { return operator=(QChar(c)); }
                     ^
/usr/include/qt5/QtCore/qstring.h:1161:28: note: in implicit copy constructor for 'QCharRef' first required here
{ Q_ASSERT(i >= 0); return QCharRef(*this, i); }
                           ^
2 warnings generated.
```

So compiling (in this example rcpconsole.cpp) with `-Wno-deprecated-copy` would silence the warning (and any other related warning, but only for the gui). Other Bitcoin Core files can be compiled normally.

hebasto commented:
@MarcoFalke mind looking into https://github.com/hebasto/bitcoin/commits/pr18738.02 ?

MarcoFalke commented:
> @MarcoFalke mind looking into https://github.com/hebasto/bitcoin/commits/pr18738.02 ?

Concept ACK

hebasto commented:
> > @MarcoFalke mind looking into https://github.com/hebasto/bitcoin/commits/pr18738.02 ?
> 
> Concept ACK

But it does not work ([comment](https://github.com/bitcoin/bitcoin/pull/18738#issuecomment-617921263)) :confused:

hebasto commented:
Updated 0025d098fc4b0f233423ed3b59dbbcd095b62ff8  -> 710a07207d484f7b6ba1e1dc35222d5964d230ee ([pr18738.01](https://github.com/hebasto/bitcoin/commits/pr18738.01) -> [pr18738.03](https://github.com/hebasto/bitcoin/commits/pr18738.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18738.01..pr18738.03)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18738#issuecomment-617888322):
> Is there a way to only add this flag when a file is compiled in the `src/qt` folder?

MarcoFalke commented on configure.ac:
```diff
@@ -355,6 +355,8 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wunused-local-typedef],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-unused-local-typedef"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wdeprecated-register],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-deprecated-register"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wimplicit-fallthrough],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-implicit-fallthrough"],,[[$CXXFLAG_WERROR]])
+  dnl Temporary workaround for QTBUG-75210 on Qt 5.12 until it is fixed.
```
This bug is also in qt 5.13.2 (not sure if it is tracked)

hebasto replied:
The patch is [merged](https://codereview.qt-project.org/c/qt/qtbase/+/245434) in Qt 5.13 tree though.

hebasto replied:
Is removing of mentioning of "Qt 5.12" will be ok?

MarcoFalke replied:
I was testing on Fedora 32, which comes with qt 5.13 and the issue is still there

hebasto replied:
> I was testing on Fedora 32, which comes with qt 5.13 and the issue is still there

Confirm.

hebasto commented:
Updated 710a07207d484f7b6ba1e1dc35222d5964d230ee -> d69174ca43e0bc8028ec2012b8787f0f6ba2f970 ([pr18738.03](https://github.com/hebasto/bitcoin/commits/pr18738.03) -> [pr18738.04](https://github.com/hebasto/bitcoin/commits/pr18738.04), [diff](https://github.com/hebasto/bitcoin/compare/pr18738.03..pr18738.04)):

- removed too narrow mention of Qt version; addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18738#discussion_r413216285):
> This bug is also in qt 5.13.2 (not sure if it is tracked)

vasild commented:
> Is there a way to only add this flag when a file is compiled in the `src/qt` folder?

Yes, we can compile `src/qt/foo.cpp` with less strict warnings. However this would be imperfect and we can do better.

Why imperfect? Because it would also suppress warnings from our code that is in `src/qt/foo.cpp`. Ideally we want to always see all warnings from our code. We may also want to see warnings from external headers, so that we are at least aware of some problem (in e.g. boost or qt), or maybe even try to fix it upstream. However we may also want to suppress warnings from external headers if they clutter the compilation output and make our warnings harder to spot, or if we want to enable `-Werror[=...]` on our code, but don't want the build to break due to warnings from external headers.

How can we do better? We can tell the compiler to suppress warnings from headers located in a particular directory:

> It seems @Empact's 422dcd5 from #14920 or @vasild's c9376ae from #18149 could help, no?

Those two commits are actually the same thing. I will submit it as a separate PR, isolated from other changes contained in the above PRs.

vasild commented:
> I will submit it as a separate PR

Here it is: https://github.com/bitcoin/bitcoin/pull/18750

fanquake commented:
I see `-Wdeprecated-copy` warnings building just `bitcoind` with LLVM Clang 10.0.0 on macOS i.e:
```bash
make src/bitcoind -j8
...
In file included from addrman.cpp:6:
In file included from ./addrman.h:15:
In file included from ./util/system.h:20:
In file included from ./fs.h:14:
In file included from /usr/local/include/boost/filesystem.hpp:16:
In file included from /usr/local/include/boost/filesystem/path.hpp:32:
/usr/local/include/boost/io/detail/quoted_manip.hpp:62:23: warning: definition of implicit copy constructor for 'quoted_proxy<const std::__1::basic_string<char> &, char>' is deprecated because it has a user-declared copy assignment operator [-Wdeprecated-copy]
        quoted_proxy& operator=(const quoted_proxy&);  // = deleted
                      ^
/usr/local/include/boost/io/detail/quoted_manip.hpp:166:14: note: in implicit copy constructor for 'boost::io::detail::quoted_proxy<const std::__1::basic_string<char> &, char>' first required here
      return detail::quoted_proxy<std::basic_string<Char, Traits, Alloc> const &, Char>
             ^
/usr/local/include/boost/filesystem/path.hpp:827:21: note: in instantiation of function template specialization 'boost::io::quoted<char, std::__1::char_traits<char>, std::__1::allocator<char> >' requested here
      << boost::io::quoted(p.template string<std::basic_string<Char> >(), static_cast<Char>('&'));
                    ^
./tinyformat.h:358:13: note: in instantiation of function template specialization 'boost::filesystem::operator<<<char, std::__1::char_traits<char> >' requested here
        out << value;
            ^
./tinyformat.h:543:13: note: in instantiation of function template specialization 'tinyformat::formatValue<boost::filesystem::path>' requested here
            formatValue(out, fmtBegin, fmtEnd, ntrunc, *static_cast<const T*>(value));
            ^
./tinyformat.h:519:27: note: in instantiation of function template specialization 'tinyformat::detail::FormatArg::formatImpl<boost::filesystem::path>' requested here
            m_formatImpl(&formatImpl<T>),
                          ^
./tinyformat.h:975:32: note: (skipping 1 context in backtrace; use -ftemplate-backtrace-limit=0 to see all)
            m_formatterStore { FormatArg(args)... }
                               ^
./tinyformat.h:1028:12: note: in instantiation of function template specialization 'tinyformat::detail::FormatListN<2>::FormatListN<boost::filesystem::path, int>' requested here
    return detail::FormatListN<sizeof...(args)>(args...);
           ^
./tinyformat.h:1064:23: note: in instantiation of function template specialization 'tinyformat::makeFormatList<boost::filesystem::path, int>' requested here
    vformat(out, fmt, makeFormatList(args...));
                      ^
./tinyformat.h:1073:5: note: in instantiation of function template specialization 'tinyformat::format<boost::filesystem::path, int>' requested here
    format(oss, fmt, args...);
    ^
./logging.h:169:28: note: in instantiation of function template specialization 'tinyformat::format<boost::filesystem::path, int>' requested here
            log_msg = tfm::format(fmt, args...);
                           ^
addrman.cpp:638:5: note: in instantiation of function template specialization 'LogPrintf<boost::filesystem::path, int>' requested here
    LogPrintf("Opened asmap file %s (%d bytes) from disk\n", path, length);
    ^
1 warning generated.
```

So if we are going to suppress these I don't think it should be qt specific.

[`-Wdeprecated-copy` warnings](https://gist.github.com/fanquake/f822aaa1d9359182a8ad9c11d0c401e0).

hebasto commented:
@fanquake I can see these `boost-1.72.0`-specific warnings too.

MarcoFalke commented:
So it seems this warning is not yet ready for general use. I guess we can disable it for 0.21.0.

Hopefully removing boost in 0.22.0 will also make the warnings go away and we can turn it on.

hebasto commented:
Reverted back d69174ca43e0bc8028ec2012b8787f0f6ba2f970  -> 0025d098fc4b0f233423ed3b59dbbcd095b62ff8 ([pr18738.04](https://github.com/hebasto/bitcoin/commits/pr18738.04) -> [pr18738.01](https://github.com/hebasto/bitcoin/commits/pr18738.01)):

- addressed:
  - @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/18738#issuecomment-622956100): 
    > So if we are going to suppress these I don't think it should be qt specific.

  - @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18738#issuecomment-623542317):
    > So it seems this warning is not yet ready for general use. I guess we can disable it for 0.21.0.

fanquake commented:
The commit message needs updating as well.

hebasto commented:
Updated 0025d098fc4b0f233423ed3b59dbbcd095b62ff8 -> 0c63f808542ba02fc41aa90b1d96e9123f16d8ad ([pr18738.01](https://github.com/hebasto/bitcoin/commits/pr18738.01) -> [pr18738.05](https://github.com/hebasto/bitcoin/commits/pr18738.05), [diff](https://github.com/hebasto/bitcoin/compare/pr18738.01..pr18738.05)):

- addressed @fanquake's [comment](hhttps://github.com/bitcoin/bitcoin/pull/18738#issuecomment-623812113):
> The commit message needs updating as well.

MarcoFalke commented:
ACK 0c63f808542ba02fc41aa90b1d96e9123f16d8ad seems fine to disable this warning for the 0.21.0 release temporarily and then enable it for 0.22.0, when boost is removed.

hebasto commented:
@MarcoFalke Is it worth backporting to 0.20 branch before rc2 for neat user experience on Ubuntu Focal?

MarcoFalke commented:
cc @fanquake ^

fanquake reviewed (approved):
ACK 0c63f808542ba02fc41aa90b1d96e9123f16d8ad - I think it's ok to suppress these for now, given that `-Wdeprecated-copy` is enabled (via `-Wextra`) in GCC 9 and Clang 10. The Qt output is pretty noisy, and there's a few warnings from Boost as well.

hebasto commented:
Backport to 0.20?
